transport/grpc/outbound_test.go

```diff
@@ -200,3 +203,159 @@ func TestCallStreamWithInvalidPeer(t *testing.T) {
+				}
+				return err
+			}),
+	)
+	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:15213"))
+	require.NoError(t, err)
+	go func() {
+		// TODO: do we have consistency issue here? Request comes before server begin to serve
+		_ = matchServer.Serve(listener)
+	}()
+	defer matchServer.Stop()
+
+	grpcTransport := NewTransport()
+	out := grpcTransport.NewSingleOutbound(listener.Addr().String())
+
+	require.NoError(t, grpcTransport.Start())
+	require.NoError(t, out.Start())
+	defer grpcTransport.Stop()
+	defer out.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*1000)
+	defer cancel()
+	req := &transport.Request{
+		Service:   "Service",
+		Procedure: "Hello",
+		Body:      bytes.NewReader([]byte("world")),
+	}
+	_, err = out.Call(ctx, req)
+	require.NoError(t, err)
+}
+
+func TestCallSvcMatchFailed(t *testing.T) {
+	mismatchServer := grpc.NewServer(
+		grpc.CustomCodec(customCodec{}),
+		grpc.UnknownServiceHandler(
+			func(srv interface{}, stream grpc.ServerStream) error {
+				responseWriter := newResponseWriter()
+				defer responseWriter.Close()
+
+				// Append a wrong service name
+				responseWriter.AddSystemHeader(ServiceHeader, "ThisIsAWrongSvcName")
+
+				if sendErr := stream.SendMsg(responseWriter.Bytes()); sendErr != nil {
+					// We couldn't send the response
+					return sendErr
+				}
+				if responseWriter.md != nil {
+					stream.SetTrailer(responseWriter.md)
+				}
+				return nil
+			}),
+	)
+	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:15214"))
+	require.NoError(t, err)
+	go func() {
+		// TODO: do we have consistency issue here? Request comes before server begin to serve
+		_ = mismatchServer.Serve(listener)
+	}()
+	defer mismatchServer.Stop()
+
+	grpcTransport := NewTransport()
+	out := grpcTransport.NewSingleOutbound(listener.Addr().String())
+
+	require.NoError(t, grpcTransport.Start())
+	require.NoError(t, out.Start())
+	defer grpcTransport.Stop()
+	defer out.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*1000)
+	defer cancel()
+	req := &transport.Request{
+		Service:   "Service",
+		Procedure: "Hello",
+		Body:      bytes.NewReader([]byte("world")),
+	}
+	_, err = out.Call(ctx, req)
+	assert.Error(t, err, "expected failure for service name dismatch")
+}
+
+func TestCallSvcMatchNoHeader(t *testing.T) {
+	noheaderServer := grpc.NewServer(
+		grpc.CustomCodec(customCodec{}),
+		grpc.UnknownServiceHandler(
+			func(srv interface{}, stream grpc.ServerStream) error {
+				responseWriter := newResponseWriter()
+				defer responseWriter.Close()
+
+				// We don't add any header or body
+
+				if sendErr := stream.SendMsg(responseWriter.Bytes()); sendErr != nil {
+					// We couldn't send the response
+					return sendErr
+				}
+				if responseWriter.md != nil {
+					stream.SetTrailer(responseWriter.md)
+				}
+				return nil
+			}),
+	)
+
+	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:15214"))
```

Comment:
Let's increment this port number by one so that all the tests are using unique ports. We've had issues in the past where another test held onto a port longer than it needed to which caused another test relying on that port number to fail.

Reply:
Resolved.